### PR TITLE
[RDY] A workaround to avoid a suspected MSVC code compilation issue

### LINKED ISSUE #17791: Protection value of clothes seems overflowed
After a long period of playing (about two hours), when I was reading the description of leggings dropped along with a zombie corpse, I suddenly found the protection values are abnormally huge, seemed they were overflowed or corrupted or some other bad things happened. Then I opened the _with for an item_ debug menu, and found all kinds of clothes in `data\json\items\armor.json`, like pants, shirts, cleats, etc., were bugged.

I restarted the program, then everything went back normal. I think there is potential problem in the source code, but I don't know how to reproduce this.

EDIT: Uploaded English screen capture

----

<del>Sorry forgot to switch language when taking the screen capture. The two values are bash protection and cut protection.</del>

### PR DESCRIPTION
**Pull request by ghost**
Fix #17791 .

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Does it actually fix the bug? It is much more ugly than the 1 line that it replaces and it doesn't look like it actually addresses the issue at all.

**Comment by DLaboratory** [comment]
After I manually expanded that complicated one-line code, variable `eff_damage` gets the correct result. But if I don't, the value of this variable is gibberish. Looks like a compiler bug.

**Comment by mugling** [comment]
> Looks like a compiler bug.
> It is much more ugly than the 1 line that it replaces

That fix isn't at all pretty. It's not impossible it's a compiler bug but that needs to be proven before this can be merged (and you should report it upstream if so). Try reducing to the smallest possible test case.

**Comment by DLaboratory** [comment]
Now I'm trying
```C++
        int eff_damage;
        if (to_self)
            eff_damage = std::min(damage(), 0);
        else
            eff_damage = std::max(damage(), 0);
```

It doesn't work.

**Comment by Coolthulhu** [comment]
Why would MSVC hate `min` or `max`? Those seem to be pretty basic.

Since those two templates work just fine in other contexts (otherwise far more things would break), this likely depends on the context.
Is it possible that MSVC inlines usage of `damage()` when it is in the same file?

Try moving the definition of `damage` function outside `item.cpp`. If that works, it means it must be context dependent, but also that the fix should change `damage` function rather than just the call sites.

**Comment by DLaboratory** [comment]
Seems this issue isn't easy to reproduce. I wrote a test programs, defined item class, copied the entire `item::bash_resist(bool)` function and completed all member functions used in `item::bash_resist(bool)` such as `item::is_armor()` and `item::get_thickness()`, and the calculation is all correct.

**Comment by DLaboratory** [comment]
A heisenbug, hmm?

**Comment by DLaboratory** [comment]
Interesting. I moved the implementation of `item::damage()` function to another cpp file, then the value is still gibberish. I turned the _whole program optimization_ off, the optimizer shouldn't be affected by another source file when processing the current.

Another information I don't know if useful: that corrupted value is always the same during one instance, if you exit and relaunch Cataclysm, it changes.

**Comment by Coolthulhu** [comment]
> Another information I don't know if useful: that corrupted value is always the same during one instance, if you exit and relaunch Cataclysm, it changes.

For different objects too?

Did you try deleting all the `obj` files and recompiling from scratch? It may be possible that MSVC caches/precompiles headers or something like that.

Try adding your fix in the `damage` function. That is, change `ceil` and `floor` into `if`/`else`.

Also, can you dump the assembly for your modified function and unmodified one somewhere?

**Comment by DLaboratory** [comment]
Ah, SSE is heavily used in the assembly.

[bugged.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/383214/bugged.txt)
[correct.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/383215/correct.txt)

**Comment by DLaboratory** [comment]
A better formatted disassembly:
[disassembly.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/383254/disassembly.zip)

**Comment by DLaboratory** [comment]
Found problem in the pseudocode.

```C
    if ( item::is_armor(v3) )
    {
      v16 = item::damage(v3);
      v17 = &v25;
      if ( v2 )
      {
        if ( v16 <= 0 )
          v17 = &v25;
      }
      else if ( v16 >= 0 )
      {
        v17 = &v25;
      }
      v18 = *(_DWORD *)v17;
      add_msg("eff_damage=  %d\n", *(_DWORD *)v17);
```
`v25` is likely an irrelevant variable, uninitialized and used in Microsoft's internal `std::_Tree` template function, has nothing to do with damage calculation, but `eff_damage` is assigned with it.

[pseudocode.txt](https://github.com/CleverRaven/Cataclysm-DDA/files/383309/pseudocode.txt)

**Comment by Coolthulhu** [comment]
It looks so weird that it probably is a bug in the implementation rather than just a quirk.

As for fixing, I see the following options:
* Extracting the "effective thickness" into a function. A smaller function may be unaffected by the bug
* Extracting the effective thickness into a function and then doing that ugly `if`/`else` inside it
* Inverting the conditional `is_armor` and checking if it works around the bug

**Comment by DLaboratory** [comment]
Now I found the pattern, if we combine `damage()` function and `std::min()` together, it causes improper assembly output, but if we manually split them, things get okay.

**Comment by DLaboratory** [comment]
Compared with the official MinGW builds, now the results are correct.


MSVC isn't a tractable compiler, :laughing:

**Comment by DLaboratory** [comment]
Sorry for my English, I actually mean _inline_ where I wrote _expand_.

**Comment by mugling** [comment]
The `pragma` fix from #17958 would be preferred

**Comment by DLaboratory** [comment]
Sorry I'm not using Windows right now, so I can't test my code.

No offence, but I personally prefer this manual expanding rather than using `#pragma` to turn off optimization. `#pragma optimize` must be outside a function, meaning that the entire `item::bash_resist` function couldn't be optimized. If the problem disappears after separate `damage()` function call and `std::min/max` apart, then I would prefer leave the optimization level on as normal.

**Comment by mugling** [comment]
`pragma` is architecturally the correct solution to a compiler bug and is both less ugly and more self-documenting as to it's intention.

**Comment by DLaboratory** [comment]
OK, modified to use that. Sorry I don't have much experience dealing with weird behaviors about compilers.

File: src/item.cpp [hunk 1]
```diff
@@ -3097,6 +3088,10 @@ int item::cut_resist( bool to_self ) const
     return lround((resist * eff_thickness * adjustment) + l_padding + k_padding);
 }
 
+#ifdef _MSC_VER
+#pragma ( "", on )
```
Comment (mugling) [inline src/item.cpp]: Is this line correct?

File: src/item.cpp [hunk 2]
```diff
@@ -2993,6 +2993,11 @@ bool item::craft_has_charges()
     return false;
 }
 
+#ifdef _MSC_VER
+// Deal with MSVC compiler bug
```
Comment (mugling) [inline src/item.cpp]: Consider referencing the GH issue in the code comment

File: src/item.cpp (hunk 1 above)
Comment (DLaboratory) [inline src/item.cpp]: Shit, I made a dumb mistake.

**Pull request merged**